hibike/devices/YogiBear/YogiBear.cpp

```diff
@@ -1,122 +1,324 @@
+    Serial.begin(115200);
+  }
+
+
+}
 
-// you must implement this function. It is called when the device receives a DeviceUpdate packet.
-// the return value is the value field of the DeviceRespond packet hibike will respond with
 /* 
-    DUTY
-    EN/DIS
-    FORWARD/REVERSE
-    FAULT LINE
-
-    future:
-    ENCODER VALUES -library?
-    PID mode vs Open Loop modes; PID Arduino library
-    Current Sense - loop update Limit the PWM if above the current for current motor
+   Manual Controls:
+   s <x> - sets pwm to x
+   c     - clears faults
+   p <x> - turns on PID mode, velocity if x = v, position if x = p
+   m     - turns on manual input mode
+   e     - enables motor
+   d     - disables motor
+   r     - displays 1 print of all readable values
+   t     - toggles continual printing of pos and vel
+   b     - send heartbeat
+   h     - switch hibike mode
+   z     - switch human controls
+   w <x> <y> - writes the value y to the variable x
 */
-uint32_t device_update(uint8_t param, uint32_t value) {
-  // if (param < NUM_PARAMS) {
-  //     params[param] = value;
-  //     return params[param];
-  //   }
-  
-  switch (param) {
-
-    case DUTY:
-      if (value <= 100) && (value >= 0) {
-        duty = value;
-      }
-      return duty;
-      break;
-      
-    case FORWARD:
-      forward = 1;
-      inputA = 0;
-      inputB = 1;
-      return forward;
-      break;
-
-    case REVERSE:
-      forward = 0;
-      inputA = 1;
-      inputB = 0;
-      return forward;
-      break;
-
-    default:
-      return ~((uint32_t) 0);
+void loop()
+{
+  ctrl_LEDs();
+  if(hibike)
+  {
+    hibike_loop();
   }
+  else
+  {
+    manual_ui();
+  }
+
 }
 
-// you must implement this function. It is called when the device receives a DeviceStatus packet.
-// the return value is the value field of the DeviceRespond packet hibike will respond with
-uint32_t device_status(uint8_t param) {
-  // if (param < NUM_PARAMS) {
-  //   return params[param];
-  // }
-  switch (param) {
-    case DUTY:
-      return duty;
+
+
+
+// You must implement this function.
+// It is called when the device receives a Device Write packet.
+// Updates param to new value passed in data.
+//    param   -   Parameter index
+//    data    -   value to write, in little-endian bytes
+//    len     -   number of bytes in data
+//
+//   return  -   size of bytes written on success; otherwise return 0
+
+uint32_t device_write(uint8_t param, uint8_t* data, size_t len) {
+  switch (param) 
+  {
+  	case ENABLE:
+	  	if (data[0] == 0) {
+	  		enabled = false;
+	  		disable();
+	  	} else {
+	  		enabled = true;
+	  		enable();
+	  	}
+      return sizeof(uint8_t);
+  	  break;
+    case COMMAND_STATE: 
+    	driveMode = data[0];
+    	if (driveMode == 0) {
+    		disablePID();
+    	} else if (driveMode == 1) {
+    		enableVel();
+    	} else {
+    		enablePos();
+    	}
+      return sizeof(driveMode);
+      break;
+    case DUTY_CYCLE: 
+    	pwmInput = ((float *)data)[0];
```

Comment:
It's interesting how everything has getters/setters other than `pwmInput`. BTW, getters/setters are exactly how you avoid having extern variables everywhere.